File: 10.regular-expression-matching.cpp
/*
 * @lc app=leetcode id=10 lang=cpp
 *
 * [10] Regular Expression Matching
 */

// @lc code=start
#include<string>
#include<vector>
using namespace std;

#define bool_or(a,b,c) ((a)||(b)||(c)) 
class Solution {
public:
    bool isMatch(string s, string p) {
        vector<vector<bool>> match(s.size()+1,vector<bool>(p.size()+1,false));
        match[s.size()][p.size()] = 1;
        for (int i = s.size(); i >= 0; i--)
        {
            for (int j = p.size(); j >= 0; j--)
            {
                if(i == s.size() && j == p.size()) continue;

                match[i][j] = bool_or(
                                       ((i < s.size() && j < p.size()) && ( p[j] == s[i] || p[j] == '.' ))?match[i+1][j+1]:false,
                                       ((i < s.size() && j < p.size()) && ( p[j+1] == '*' && (p[j]== s[i] || p[j] == '.') ))?match[i+1][j]:false,
                                       ((j < p.size()-1) && p[j+1] == '*')?match[i][j+2]:false
                                     );
            }
        }
        return match[0][0];
    }

    // void print(vector<vector<bool>> &match){
    //     for (size_t i = 0; i < match.size(); i++)
    //     {
    //         for (size_t j = 0; j < match[0].size(); j++)
    //         {
    //             cout<<match[i][j]<<'|';
    //         }
    //         cout<<endl;
    //         /* code */
    //     }
    //     cout<<endl<<"============="<<endl;
    // }
};
// @lc code=end

File: 10.regular-expression-matching.go
/*
 * @lc app=leetcode id=10 lang=golang
 *
 * [10] Regular Expression Matching
 */

// @lc code=start
type State struct {
	i, j int
}

var memo map[State]bool

var _s, _p string

func isMatch(s string, p string) bool {
	_s, _p = s, p

	memo = make(map[State]bool)
	return dp(0, 0)
}

func dp(i, j int) (matched bool) {
	state := State{i: i, j: j}
	if v, ok := memo[state]; ok {
		return v
	}

	if i >= len(_s) && j >= len(_p) {
		memo[state] = true
		return memo[state]
	} else if j >= len(_p) {
		memo[state] = false
		return memo[state]
	} else if i == len(_s) {
		// expect p[j: len(p)] can equal to ""...
		// like a*b*c*d*, all the *'s elimate previous letter
		if j < len(_p)-1 && _p[j+1] == '*' {
			memo[state] = dp(i, j+2)
			return memo[state]
		} else {
			memo[state] = false
			return memo[state]
		}
	}

	matched = false
	if j < len(_p)-1 && _p[j+1] == '*' {
		if _s[i] == _p[j] || _p[j] == '.' {
			matched = dp(i+1, j) || dp(i+1, j+2) || dp(i, j+2)
		} else {
			matched = dp(i, j+2)
		}
	} else if _s[i] == _p[j] || _p[j] == '.' {
		matched = dp(i+1, j+1)
	}

	memo[state] = matched
	return memo[state]
}

// @lc code=end
